### PR TITLE
Use copyToPointer in encodeUtf8Builder

### PR DESCRIPTION
**Pull request by alexbiehl**
Instead of copying each character separately by using `encodeUtf8BuilderEscaped` we instead resort to memcpy to copy the bytes into the target buffer more efficiently.

This PR is based off similar functions in the ByteString package: 

https://github.com/haskell/bytestring/blob/ba3046d9f5bf5cac1246c39f519b8be7ac11f49c/Data/ByteString/Builder/Internal.hs#L832-L845

https://github.com/haskell/bytestring/blob/ba3046d9f5bf5cac1246c39f519b8be7ac11f49c/Data/ByteString/Builder/Internal.hs#L779-L796

### CODE REVIEW

**Comment by Boarders** [comment]
Did you benchmark the change by any chance?

**Comment by alexbiehl** [comment]
@Boarders No, I did not. But shouldn't be hard to set something up I guess?

**Comment by Boarders** [comment]
@alexbiehl I'm sure it is better but always good to confirm.

**Comment by Bodigrim** [comment]
It appears we do not have any tests for `encodeUtf8Builder`. @alexbiehl could you possibly add some?

**Comment by alexbiehl** [comment]
I set up a small benchmark for encodeUtf8Builder: 

```
All
  EncodeUtf8
    non-ASCII
      Text:                          OK (2.19s)
        251  μs ±  23 μs
      LazyText:                      OK (3.14s)
        190  μs ±  16 μs
      Text/encodeUtf8Builder:        OK (20.15s)
        298  ns ±  13 ns
      Text/encodeUtf8BuilderEscaped: OK (2.07s)
        3.91 μs ± 213 ns
    ASCII
      Text:                          OK (1.34s)
        78.2 μs ± 7.2 μs
      LazyText:                      OK (8.72s)
        65.7 μs ± 1.8 μs
      Text/encodeUtf8Builder:        OK (1.30s)
        293  ns ±  27 ns
      Text/encodeUtf8BuilderEscaped: OK (2.28s)
        8.67 μs ± 462 ns

All 8 tests passed (41.26s)
```

File: benchmarks/haskell/Benchmarks/EncodeUtf8.hs [hunk 1]
```diff
@@ -23,6 +25,8 @@ benchmark name string =
     bgroup name
         [ bench "Text"     $ whnf (B.length . T.encodeUtf8)   text
         , bench "LazyText" $ whnf (BL.length . TL.encodeUtf8) lazyText
+        , bench "Text/encodeUtf8Builder" $ whnf (B.toLazyByteString . T.encodeUtf8Builder) text
+        , bench "Text/encodeUtf8BuilderEscaped" $ whnf (B.toLazyByteString . T.encodeUtf8BuilderEscaped (BP.liftFixedToBounded BP.word8)) text
```
Comment (Bodigrim) [inline benchmarks/haskell/Benchmarks/EncodeUtf8.hs:29]: This looks a bit weird: `whnf` of a lazy bytestring does not trigger anything except the first chunk.

File: tests/Tests/Properties/Transcoding.hs [hunk 2]
```diff
@@ -49,6 +52,11 @@ tl_utf32LE   = (EL.decodeUtf32LE . EL.encodeUtf32LE) `eq` id
 t_utf32BE    = (E.decodeUtf32BE . E.encodeUtf32BE) `eq` id
 tl_utf32BE   = (EL.decodeUtf32BE . EL.encodeUtf32BE) `eq` id
 
+t_encodeUtf8Builder = (E.decodeUtf8 . BL.toStrict . B.toLazyByteString . E.encodeUtf8Builder) `eq` id
+t_encodeUtf8Builder_sanity t =
+  (B.toLazyByteString . E.encodeUtf8Builder) t ==
```
Comment (Bodigrim) [inline tests/Tests/Properties/Transcoding.hs:57]:
```suggestion
  (B.toLazyByteString . E.encodeUtf8Builder) t ===
```

File: benchmarks/haskell/Benchmarks/EncodeUtf8.hs (hunk 1 above)
Comment (alexbiehl) [reply benchmarks/haskell/Benchmarks/EncodeUtf8.hs:29]: Ah, right!

**Comment by alexbiehl** [comment]
Much more realistic now: 

```
All
  EncodeUtf8
    non-ASCII
      Text:                          OK (2.24s)
        262  μs ±  15 μs
      LazyText:                      OK (1.78s)
        208  μs ±  19 μs
      Text/encodeUtf8Builder:        OK (9.07s)
        272  μs ± 6.2 μs
      Text/encodeUtf8BuilderEscaped: OK (1.53s)
        2.94 ms ± 247 μs
    ASCII
      Text:                          OK (1.41s)
        81.7 μs ± 6.8 μs
      LazyText:                      OK (4.64s)
        69.3 μs ± 3.7 μs
      Text/encodeUtf8Builder:        OK (2.89s)
        86.5 μs ± 6.3 μs
      Text/encodeUtf8BuilderEscaped: OK (2.41s)
        2.33 ms ± 154 μs

```

**Comment by Boarders** [comment]
@alexbiehl Sorry to be a pain but could you run the benchmark on the previous implementation just to see how this changes things?

**Comment by alexbiehl** [comment]
@Boarders You can tell by looking at the `Text/encodeUtf8BuilderEscaped` line.

**Comment by Boarders** [comment]
Looks good, thank you very much for adding it!

File: src/Data/Text/Encoding.hs [hunk 3]
```diff
@@ -391,7 +391,37 @@ decodeUtf8Lenient = decodeUtf8With lenientDecode
...
+          unsafeSTToIO $ A.copyToPointer arr off op inpRemaining
+          let !br = B.BufferRange (op `plusPtr` inpRemaining) ope
+          k br
+      | otherwise = do
+          unsafeSTToIO $ A.copyToPointer arr off op outRemaining
+          let !off' = off + outRemaining
+          return $ B.bufferFull 1 ope (go off')
+      where
+        outRemaining = ope `minusPtr` op
+        inpRemaining = len - off
```
Comment (Bodigrim) [inline src/Data/Text/Encoding.hs:424]: @alexbiehl could you please add a test, which exercises `encodeUtf8Builder` with non-zero `off`?

Comment (alexbiehl) [reply src/Data/Text/Encoding.hs:424]: Is there any existing test that does so? No idea how to test that effectively.

Comment (alexbiehl) [reply src/Data/Text/Encoding.hs]: Ah, got it.

Comment (alexbiehl) [reply src/Data/Text/Encoding.hs]: And indeed it catched the bug!

File: src/Data/Text/Encoding.hs [hunk 4]
```diff
@@ -391,7 +391,37 @@ decodeUtf8Lenient = decodeUtf8With lenientDecode
...
+  where
+    go !ip !ipe (B.BufferRange op ope)
+      | inpRemaining <= outRemaining = do
+          unsafeSTToIO $ A.copyToPointer arr ip op inpRemaining
+          let !br = B.BufferRange (op `plusPtr` inpRemaining) ope
+          k br
+      | otherwise = do
+          unsafeSTToIO $ A.copyToPointer arr ip op outRemaining
+          let !ip' = ip + outRemaining
+          return $ B.bufferFull 1 ope (go ip' ipe)
```
Comment (Bodigrim) [inline src/Data/Text/Encoding.hs:421]: It could be that I terribly misunderstand how `bufferFull` works, but why do we pass `1` instead of `inpRemaining - outRemaining`? We could have avoided any recursive functions at all, no?

Comment (alexbiehl) [reply src/Data/Text/Encoding.hs:421]: I think it signals „I need at least one byte to make progress“. As stated in the description this is copied more or less from the ByteString and ShortByteString cases.

Comment (alexbiehl) [reply src/Data/Text/Encoding.hs:421]: Comment on bufferFull arg says 
„Minimal size of next 'BufferRange'.“

Comment (Bodigrim) [reply src/Data/Text/Encoding.hs:421]: Yeah, I have the same question about implementations in `bytestring` :)

Comment (alexbiehl) [reply src/Data/Text/Encoding.hs:421]: So to me it looks like conservative default. If you’d pass the the length of the remaining string you’d put assumptions on the upper bound of the size of the buffer you are writing to while we really want the driver to be in charge of the buffer alone. I could def. see a compromise of passing upper and lower bound together so that the driver can decide how much to allocate more precisely.

**Review by Bodigrim — APPROVED** [review]

**Review by Lysxia — APPROVED** [review]
